File: ios/Classes/InMobiAdsPlugin.swift
import Flutter
import InMobiSDK
import UIKit

/// Bridges the InMobi iOS SDK onto one method channel.
///
/// The iOS SDK binds its delegate to the ad object rather than to the request,
/// so ads are held in `fullScreenAds` against the id Dart minted, and every
/// event carries that id back. `AdDelegate` is a separate object per ad for the
/// same reason — the delegate callbacks do not identify which request they
/// belong to.
public class InMobiAdsPlugin: NSObject, FlutterPlugin {

    private let channel: FlutterMethodChannel
    private var fullScreenAds: [Int: IMInterstitial] = [:]
    private var delegates: [Int: FullScreenAdDelegate] = [:]

    init(channel: FlutterMethodChannel) {
        self.channel = channel
        super.init()
    }

    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(
            name: "inmobi_ads",
            binaryMessenger: registrar.messenger()
        )
        let instance = InMobiAdsPlugin(channel: channel)
        registrar.addMethodCallDelegate(instance, channel: channel)

        registrar.register(
            InMobiBannerViewFactory(messenger: registrar.messenger()) {
                [weak instance] adId, event, arguments in
                instance?.sendEvent(adId: adId, event: event, arguments: arguments)
            },
            withId: "inmobi_ads/banner"
        )
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        let arguments = call.arguments as? [String: Any] ?? [:]

        switch call.method {
        case "initialize":
            initialize(arguments, result)
        case "setConsent":
            IMSdk.updateGDPRConsent(Self.consentDictionary(arguments))
            result(nil)
        case "setLogLevel":
            IMSdk.setLogLevel(Self.logLevel(arguments["logLevel"] as? String))
            result(nil)
        case "loadFullScreenAd":
            loadFullScreenAd(arguments, result)
        case "showFullScreenAd":
            showFullScreenAd(arguments, result)
        case "disposeAd":
            if let adId = arguments["adId"] as? Int {
                fullScreenAds.removeValue(forKey: adId)
                delegates.removeValue(forKey: adId)
            }
            result(nil)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private func initialize(_ arguments: [String: Any], _ result: @escaping FlutterResult) {
        guard let accountId = arguments["accountId"] as? String, !accountId.isEmpty else {
            result(FlutterError(
                code: "INVALID_ACCOUNT_ID",
                message: "accountId must not be empty",
                details: nil
            ))
            return
        }

        IMSdk.setLogLevel(Self.logLevel(arguments["logLevel"] as? String))

        let consent = Self.consentDictionary(arguments["consent"] as? [String: Any] ?? [:])
        IMSdk.initWithAccountID(accountId, consentDictionary: consent) { error in
            if let error = error {
                result(FlutterError(
                    code: "INIT_FAILED",
                    message: error.localizedDescription,
                    details: nil
                ))
            } else {
                result(nil)
            }
        }
    }

    private func loadFullScreenAd(_ arguments: [String: Any], _ result: @escaping FlutterResult) {
        guard
            let adId = arguments["adId"] as? Int,
            let placementId = (arguments["placementId"] as? NSNumber)?.int64Value
        else {
            result(FlutterError(
                code: "INVALID_ARGUMENTS",
                message: "adId and placementId are required",
                details: nil
            ))
            return
        }

        let delegate = FullScreenAdDelegate(adId: adId) { [weak self] adId, event, arguments in
            self?.sendEvent(adId: adId, event: event, arguments: arguments)
        }
        let ad = IMInterstitial(placementId: placementId)
        ad.delegate = delegate

        // The delegate is held here rather than by the ad, because IMInterstitial
        // keeps its delegate weakly — letting it go out of scope means the ad
        // loads and no callback ever arrives.
        delegates[adId] = delegate
        fullScreenAds[adId] = ad

        ad.load()
        result(nil)
    }

    private func showFullScreenAd(_ arguments: [String: Any], _ result: @escaping FlutterResult) {
        guard let adId = arguments["adId"] as? Int else {
            result(FlutterError(
                code: "INVALID_ARGUMENTS",
                message: "adId is required",
                details: nil
            ))
            return
        }

        guard let ad = fullScreenAds[adId] else {
            sendEvent(
                adId: adId,
                event: "displayFailed",
                arguments: ["code": "INTERNAL_ERROR", "message": "Ad \(adId) is gone"]
            )
            result(nil)
            return
        }

        guard let presenter = Self.topViewController() else {
            sendEvent(
                adId: adId,
                event: "displayFailed",
                arguments: [
                    "code": "INTERNAL_ERROR",
                    "message": "No view controller available to present from",
                ]
            )
            result(nil)
            return
        }

        // isReady is checked here because show() on a spent or unloaded
        // IMInterstitial fails silently — no delegate callback at all — which
        // would leave the caller's await hanging until its own timeout.
        guard ad.isReady() else {
            sendEvent(
                adId: adId,
                event: "displayFailed",
                arguments: ["code": "INTERNAL_ERROR", "message": "Ad \(adId) is not ready"]
            )
            result(nil)
            return
        }

        ad.show(from: presenter)
        result(nil)
    }

    private func sendEvent(adId: Int, event: String, arguments: [String: Any?]) {
        var payload = arguments
        payload["adId"] = adId
        payload["event"] = event
        DispatchQueue.main.async { [weak self] in
            self?.channel.invokeMethod("onAdEvent", arguments: payload)
        }
    }

    /// The frontmost view controller, walking past anything already presented.
    ///
    /// Presenting from a controller that is itself covered throws a UIKit
    /// warning and shows nothing, which is easy to hit when an ad is triggered
    /// from a Flutter route shown over a native modal.
    private static func topViewController() -> UIViewController? {
        let root = UIApplication.shared.connectedScenes
            .compactMap { $0 as? UIWindowScene }
            .flatMap { $0.windows }
            .first { $0.isKeyWindow }?
            .rootViewController

        var top = root
        while let presented = top?.presentedViewController {
            top = presented
        }
        return top
    }

    private static func logLevel(_ name: String?) -> IMSDKLogLevel {
        switch name {
        case "error": return .error
        case "debug": return .debug
        default: return .none
        }
    }

    /// Translates this package's normalised consent model into InMobi's
    /// dictionary. The key names live here, and only here.
    private static func consentDictionary(_ consent: [String: Any]) -> [String: Any] {
        var dictionary: [String: Any] = [:]
        let gdprApplies = consent["gdprApplies"] as? Bool ?? false
        dictionary["gdpr"] = gdprApplies ? "1" : "0"

        if let consentString = consent["consentString"] as? String {
            dictionary["gdpr_consent"] = consentString
        }
        if let consentGiven = consent["consentGiven"] as? Bool {
            dictionary["gdpr_consent_available"] = consentGiven
        }
        return dictionary
    }
}

/// One ad's delegate, tagged with the id its events belong to.
final class FullScreenAdDelegate: NSObject, IMInterstitialDelegate {

    private let adId: Int
    private let sendEvent: (Int, String, [String: Any?]) -> Void

    init(adId: Int, sendEvent: @escaping (Int, String, [String: Any?]) -> Void) {
        self.adId = adId
        self.sendEvent = sendEvent
        super.init()
    }

    func interstitialDidFinishLoading(_ interstitial: IMInterstitial) {
        sendEvent(adId, "loaded", [:])
    }

    func interstitial(_ interstitial: IMInterstitial, didFailToLoadWithError error: IMRequestStatus) {
        sendEvent(adId, "loadFailed", InMobiStatus.eventMap(error))
    }

    func interstitialDidPresent(_ interstitial: IMInterstitial) {
        sendEvent(adId, "displayed", [:])
    }

    func interstitial(_ interstitial: IMInterstitial, didFailToPresentWithError error: IMRequestStatus) {
        sendEvent(adId, "displayFailed", InMobiStatus.eventMap(error))
    }

    func interstitialAdImpressed(_ interstitial: IMInterstitial) {
        sendEvent(adId, "impression", [:])
    }

    func interstitial(_ interstitial: IMInterstitial, didInteractWithParams params: [String: Any]?) {
        sendEvent(adId, "clicked", [:])
    }

    /// Fires before `interstitialDidDismiss`, which is what lets a caller treat
    /// dismissal as terminal and still know whether a reward was earned.
    func interstitial(_ interstitial: IMInterstitial, rewardActionCompletedWithRewards rewards: [String: Any]) {
        sendEvent(adId, "rewards", ["rewards": rewards])
    }

    func interstitialDidDismiss(_ interstitial: IMInterstitial) {
        sendEvent(adId, "dismissed", [:])
    }
}

/// Maps InMobi's iOS status codes onto the Android spelling.
///
/// The two SDKs report the same conditions under different names. Normalising
/// here means `InMobiAdError.code` means one thing in Dart, and a caller can
/// switch on `NO_FILL` without asking which platform it is on.
enum InMobiStatus {
    static func eventMap(_ status: IMRequestStatus) -> [String: Any?] {
        return [
            "code": name(for: status.code),
            "message": status.localizedDescription,
        ]
    }

    private static func name(for code: IMStatusCode) -> String {
        switch code {
        case .noFill: return "NO_FILL"
        case .networkUnReachable: return "NETWORK_UNREACHABLE"
        case .requestTimedOut: return "REQUEST_TIMED_OUT"
        case .requestInvalid: return "REQUEST_INVALID"
        case .requestPending: return "REQUEST_PENDING"
        case .serverError: return "SERVER_ERROR"
        case .internalError: return "INTERNAL_ERROR"
        case .adActive: return "AD_ACTIVE"
        case .earlyRefreshRequest: return "EARLY_REFRESH_REQUEST"
        case .monetizationDisabled: return "MONETIZATION_DISABLED"
        case .lowMemory: return "LOW_MEMORY"
        @unknown default: return "INTERNAL_ERROR"
        }
    }
}

File: android/src/main/kotlin/com/illuminationdevelopment/inmobi_ads/InMobiAdsPlugin.kt
package com.illuminationdevelopment.inmobi_ads

import android.app.Activity
import android.content.Context
import android.os.Handler
import android.os.Looper
import com.inmobi.ads.AdMetaInfo
import com.inmobi.ads.InMobiAdRequestStatus
import com.inmobi.ads.InMobiInterstitial
import com.inmobi.ads.listeners.InterstitialAdEventListener
import com.inmobi.sdk.InMobiSdk
import com.inmobi.sdk.SdkInitializationListener
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import org.json.JSONObject

/**
 * Bridges the InMobi Android SDK onto one method channel.
 *
 * Two constraints from the SDK shape everything here:
 *
 *  - `InMobiInterstitial` is **not thread-safe and must be constructed and
 *    called on the UI thread**. Every entry point below hops to the main
 *    looper rather than trusting the caller, because Flutter guarantees method
 *    calls arrive on the platform thread but says nothing about which thread
 *    that is on every embedding.
 *  - Its listener is bound to the ad instance, not the request, so ads are held
 *    in [fullScreenAds] against the id Dart minted and events carry that id back.
 */
class InMobiAdsPlugin : FlutterPlugin, ActivityAware, MethodChannel.MethodCallHandler {

    private lateinit var channel: MethodChannel
    private lateinit var applicationContext: Context
    private var activity: Activity? = null

    private val fullScreenAds = mutableMapOf<Int, InMobiInterstitial>()
    private val mainHandler = Handler(Looper.getMainLooper())

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        applicationContext = binding.applicationContext
        channel = MethodChannel(binding.binaryMessenger, CHANNEL_NAME)
        channel.setMethodCallHandler(this)

        binding.platformViewRegistry.registerViewFactory(
            BANNER_VIEW_TYPE,
            InMobiBannerViewFactory { adId, event, arguments ->
                sendEvent(adId, event, arguments)
            },
        )
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
        mainHandler.post {
            fullScreenAds.clear()
        }
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        activity = binding.activity
    }

    override fun onDetachedFromActivity() {
        activity = null
    }

    override fun onDetachedFromActivityForConfigChanges() {
        activity = null
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "initialize" -> initialize(call, result)
            "setConsent" -> {
                InMobiSdk.updateGDPRConsent(consentJson(call.arguments as? Map<*, *>))
                result.success(null)
            }
            "setLogLevel" -> {
                InMobiSdk.setLogLevel(logLevel(call.argument<String>("logLevel")))
                result.success(null)
            }
            "loadFullScreenAd" -> loadFullScreenAd(call, result)
            "showFullScreenAd" -> showFullScreenAd(call, result)
            "disposeAd" -> {
                val adId = call.argument<Int>("adId")
                mainHandler.post { adId?.let(fullScreenAds::remove) }
                result.success(null)
            }
            else -> result.notImplemented()
        }
    }

    private fun initialize(call: MethodCall, result: MethodChannel.Result) {
        val accountId = call.argument<String>("accountId")
        if (accountId.isNullOrBlank()) {
            result.error("INVALID_ACCOUNT_ID", "accountId must not be empty", null)
            return
        }

        InMobiSdk.setLogLevel(logLevel(call.argument<String>("logLevel")))

        mainHandler.post {
            InMobiSdk.init(
                applicationContext,
                accountId,
                consentJson(call.argument<Map<*, *>>("consent")),
                object : SdkInitializationListener {
                    override fun onInitializationComplete(error: Error?) {
                        if (error == null) {
                            result.success(null)
                        } else {
                            result.error(
                                "INIT_FAILED",
                                error.message ?: "InMobi SDK failed to initialize",
                                null,
                            )
                        }
                    }
                },
            )
        }
    }

    private fun loadFullScreenAd(call: MethodCall, result: MethodChannel.Result) {
        val adId = call.argument<Int>("adId")
        // Dart sends this as a 64-bit int; MethodChannel narrows small values to
        // Integer, so read it as Number and widen rather than casting to Long.
        val placementId = (call.argument<Any>("placementId") as? Number)?.toLong()

        if (adId == null || placementId == null) {
            result.error("INVALID_ARGUMENTS", "adId and placementId are required", null)
            return
        }

        // InMobi needs an Activity to present a full-screen ad. Failing here
        // rather than at show() means the caller learns about it while it can
        // still fall through to another network.
        val context = activity ?: run {
            result.error(
                "NO_ACTIVITY",
                "InMobi full-screen ads need a foreground Activity; none is attached",
                null,
            )
            return
        }

        mainHandler.post {
            val ad = InMobiInterstitial(context, placementId, FullScreenListener(adId))
            fullScreenAds[adId] = ad
            ad.load()
        }
        result.success(null)
    }

    private fun showFullScreenAd(call: MethodCall, result: MethodChannel.Result) {
        val adId = call.argument<Int>("adId")
        if (adId == null) {
            result.error("INVALID_ARGUMENTS", "adId is required", null)
            return
        }

        mainHandler.post {
            val ad = fullScreenAds[adId]
            when {
                ad == null -> sendEvent(
                    adId,
                    "displayFailed",
                    mapOf("code" to "INTERNAL_ERROR", "message" to "Ad $adId is gone"),
                )
                !ad.isReady -> sendEvent(
                    adId,
                    "displayFailed",
                    mapOf("code" to "INTERNAL_ERROR", "message" to "Ad $adId is not ready"),
                )
                else -> ad.show()
            }
        }
        result.success(null)
    }

    /** Bridges one ad's listener callbacks onto [sendEvent]. */
    private inner class FullScreenListener(private val adId: Int) :
        InterstitialAdEventListener() {

        override fun onAdLoadSucceeded(ad: InMobiInterstitial, info: AdMetaInfo) =
            sendEvent(adId, "loaded")

        override fun onAdLoadFailed(ad: InMobiInterstitial, status: InMobiAdRequestStatus) =
            sendEvent(adId, "loadFailed", status.toEventMap())

        override fun onAdDisplayed(ad: InMobiInterstitial, info: AdMetaInfo) =
            sendEvent(adId, "displayed")

        override fun onAdDisplayFailed(ad: InMobiInterstitial) = sendEvent(
            adId,
            "displayFailed",
            mapOf(
                "code" to "INTERNAL_ERROR",
                "message" to "InMobi reported onAdDisplayFailed",
            ),
        )

        override fun onAdImpression(ad: InMobiInterstitial) = sendEvent(adId, "impression")

        override fun onAdClicked(ad: InMobiInterstitial, params: MutableMap<Any, Any>?) =
            sendEvent(adId, "clicked")

        /**
         * Fires before [onAdDismissed], which is what lets a caller treat
         * dismissal as terminal and still know whether a reward was earned.
         */
        override fun onRewardsUnlocked(ad: InMobiInterstitial, rewards: MutableMap<Any, Any>) =
            sendEvent(
                adId,
                "rewards",
                mapOf("rewards" to rewards.entries.associate { it.key.toString() to it.value }),
            )

        override fun onAdDismissed(ad: InMobiInterstitial) = sendEvent(adId, "dismissed")
    }

    private fun sendEvent(
        adId: Int,
        event: String,
        arguments: Map<String, Any?> = emptyMap(),
    ) {
        val payload = HashMap<String, Any?>(arguments).apply {
            put("adId", adId)
            put("event", event)
        }
        mainHandler.post { channel.invokeMethod("onAdEvent", payload) }
    }

    private companion object {
        const val CHANNEL_NAME = "inmobi_ads"
        const val BANNER_VIEW_TYPE = "inmobi_ads/banner"

        fun logLevel(name: String?): InMobiSdk.LogLevel = when (name) {
            "error" -> InMobiSdk.LogLevel.ERROR
            "debug" -> InMobiSdk.LogLevel.DEBUG
            else -> InMobiSdk.LogLevel.NONE
        }

        /**
         * Translates this package's normalised consent model into InMobi's JSON.
         *
         * The key names live here, and only here, because InMobi has renamed
         * them across SDK releases and the Dart API should not have to move
         * when they do.
         */
        fun consentJson(consent: Map<*, *>?): JSONObject {
            val json = JSONObject()
            if (consent == null) return json

            val gdprApplies = consent["gdprApplies"] as? Boolean ?: false
            json.put("gdpr", if (gdprApplies) "1" else "0")

            (consent["consentString"] as? String)?.let { json.put("gdpr_consent", it) }
            (consent["consentGiven"] as? Boolean)?.let {
                json.put("gdpr_consent_available", it)
            }
            return json
        }
    }
}
